### PR TITLE
Kill callback child processes when it is necessary

### PR DESCRIPTION
**Pull request by CyberDem0n**
Not doing so makes it hard to implement callbacks in bash and eventually can lead to the situation when two callbacks are running at the same time. In case if we failed to kill the child process we will still wait for it to finish.

The same problem could happen with custom bootstrap, therefore we happen to kill the custom bootstrap process we also kill all child subprocesses.

Closes https://github.com/zalando/patroni/issues/1238

### CODE REVIEW

File: patroni/postgresql/callback_executor.py [hunk 1]
```diff
@@ -13,28 +14,59 @@ def __init__(self):
...
+        self._callback_event.set()
+
+    def _cancel(self):
         with self._lock:
-            if self._process and self._process.poll() is None:
+            if self._process and self._process.poll() is None and not self._process_children:
+                try:
+                    self._process_children = self._process.children(recursive=True)
+                except psutil.Error:
+                    pass
```
Comment (sdudoladov) [inline patroni/postgresql/callback_executor.py]: please at least log the error

File: patroni/postgresql/callback_executor.py [hunk 2]
```diff
@@ -13,28 +14,59 @@ def __init__(self):
...
                 try:
                     self._process.kill()
-                    logger.warning('Killed the old callback process because it was still running: %s', self._cmd)
-                except OSError:
+                    logger.warning('Killed the old callback process because it was still running: %s',
+                                   self._process_cmd)
+                except psutil.NoSuchProcess:
+                    return
+                except psutil.AccessDenied:
                     logger.exception('Failed to kill the old callback')
```
Comment (sdudoladov) [inline patroni/postgresql/callback_executor.py]: sth like 'Failed to kill the old callback (access denied)'  would imo be  clearer in the logs

File: patroni/postgresql/callback_executor.py [hunk 3]
```diff
@@ -13,28 +14,59 @@ def __init__(self):
...
+
+    def _wait(self):
+        self._process.wait()
+
+        waitlist = []
+        with self._lock:
+            for child in self._process_children:
+                if child.is_running():
+                    try:
+                        child.kill()
```
Comment (sdudoladov) [inline patroni/postgresql/callback_executor.py]: why directly kill and not terminate followed by a kill later ? that can (at least in theory) enable callback children to terminate gracefully

Comment (CyberDem0n) [reply patroni/postgresql/callback_executor.py]: We really want to kill them, as fast as possible.

File: patroni/postgresql/cancellable.py [hunk 4]
```diff
@@ -10,13 +11,66 @@
...
+            self._process = psutil.Popen(cmd, *args, **kwargs)
+        except Exception:
+            return logger.exception('Failed to execute %s',  cmd)
+        return True
+
+    def _kill_process(self):
+        with self._lock:
+            if self._process is not None and self._process.is_running() and not self._process_children:
+                try:
+                    self._process.suspend()  # Suspend the process before getting list of childrens
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/postgresql/cancellable.py]: The callback process might spawn childs. If it is suspended, it won't be able to do that.
Our goal to make sure that we killed not only the callback process, but also all children.

**Comment by sdudoladov** [comment]
:+1:

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**